### PR TITLE
Initial contract as swagger.json

### PR DESCRIPTION
**Pull request by MarkusTiede**
Currently the [Swagger.IO-Editor](https://editor.swagger.io/) lists the errors as infos

image

### CODE REVIEW

File: docs/swagger.json [hunk 1]
```diff
@@ -0,0 +1,651 @@
...
+				"address": {
+					"$ref": "#/definitions/Address"
+				}
+			}
+		},
+		"Contract": {
+			"type": "object",
+			"required": [
+				"agent",
+				"companyInvoiceAddress",
```
Comment (matthewwardle) [inline docs/swagger.json]: Should this not only be required if payment type is invoice and payer = company?

File: docs/swagger.json [hunk 2]
```diff
@@ -0,0 +1,651 @@
...
+				"discountCode",
```
Comment (matthewwardle) [inline docs/swagger.json]: Discount code should not be required, it's optional

File: docs/swagger.json [hunk 3]
```diff
@@ -0,0 +1,651 @@
...
+				"distributor",
+				"endDate",
+				"id",
+				"personInvoiceAddress",
```
Comment (matthewwardle) [inline docs/swagger.json]: Should this not only be required if payment type is invoice and payer = person?

File: docs/swagger.json [hunk 4]
```diff
@@ -0,0 +1,651 @@
...
+				"policyId",
+				"policyIssueDate",
+				"product",
+				"roles",
+				"stampTax",
+				"startDate",
+				"subAgent",
+				"tcsNumber",
```
Comment (matthewwardle) [inline docs/swagger.json]: TCS number is only on TCS policies and even then not required.

File: docs/swagger.json [hunk 5]
```diff
@@ -0,0 +1,651 @@
...
+		},
+		"Contract": {
+			"type": "object",
+			"required": [
+				"agent",
+				"companyInvoiceAddress",
+				"discountCode",
+				"distributor",
+				"endDate",
+				"id",
```
Comment (matthewwardle) [inline docs/swagger.json]: What is this ID?

File: docs/swagger.json (hunk 3 above)
Comment (matthewwardle) [reply docs/swagger.json]: Also do you want to know payment method? invoice, stripe PayPal?

File: docs/swagger.json [hunk 6]
```diff
@@ -0,0 +1,651 @@
...
+					"200": {
+						"description": "successful operation",
+						"schema": {
+							"type": "string"
+						}
+					}
+				}
+			}
+		},
+		"/documents": {
```
Comment (matthewwardle) [inline docs/swagger.json]: Can we clarify here the order of operations, as it's unclear.

In previous discussions with Frank it was.

1.  Upload document, which returns a token
2. Make policy request, including token for document so you know what to copy in background.

Here, it looks like

1. Upload policy?
2. Upload documents with policy Id reference. 

Could you clarify?  I think at least we should be saying in policy what document as expected

File: docs/swagger.json [hunk 7]
```diff
@@ -0,0 +1,651 @@
...
+				"agent",
+				"companyInvoiceAddress",
+				"discountCode",
+				"distributor",
+				"endDate",
+				"id",
+				"personInvoiceAddress",
+				"policyId",
+				"policyIssueDate",
+				"product",
```
Comment (matthewwardle) [inline docs/swagger.json]: There are a lot of fields here?  Perhaps you could provide the CSV we send you, with each field here referenced, so we have mapping table?

File: docs/swagger.json [hunk 8]
```diff
@@ -0,0 +1,651 @@
...
+			"required": [
+				"address",
+				"legalForm",
+				"name",
+				"uidNumber"
+			],
+			"properties": {
+				"name": {
+					"type": "string",
+					"minLength": 0,
```
Comment (matthewwardle) [inline docs/swagger.json]: We have this as 3

File: docs/swagger.json [hunk 9]
```diff
@@ -0,0 +1,651 @@
...
+					"maxLength": 100
```
Comment (matthewwardle) [inline docs/swagger.json]: We have been told max 70

Comment (StephanSST) [reply docs/swagger.json]: There are going to be a lot of changes. Look at this as a first try, maybe a base for discussions. We needed a contract for our firewall/security team to open up a gate for you. Hopefully next week we can provide a better version with documentation tags, correct field lenghts and optimized inheritance. Certainly we will include your feedback, thank you very much!

Comment (matthewwardle) [reply docs/swagger.json]: Cool. Thanks.   I think the most pressing question for us is the one about this.

Can we clarify here the order of operations, as it's unclear.

In previous discussions with Frank it was.

Upload document, which returns a token
Make policy request, including token for document so you know what to copy in background.
Here, it looks like

Upload policy?
Upload documents with policy Id reference.
Could you clarify? I think at least we should be saying in policy what document as expected

As we want to start looking at the integration on our end

File: docs/swagger.json (hunk 2 above)
Comment (frank-baier) [reply docs/swagger.json]: accepted

File: docs/swagger.json (hunk 1 above)
Comment (frank-baier) [reply docs/swagger.json]: accepted

File: docs/swagger.json (hunk 3 above)
Comment (frank-baier) [reply docs/swagger.json]: accepted. the code is the payment method - e.g. 12035 - Vorkasse

File: docs/swagger.json (hunk 4 above)
Comment (frank-baier) [reply docs/swagger.json]: accepted - moved to Person and renamed to membershipId to be more open for other organizations

File: docs/swagger.json (hunk 5 above)
Comment (frank-baier) [reply docs/swagger.json]: we thought of an ID to identify a request uniquely for idempotency reasons. Removed it from Contract. We think about adding a requestId as second Param to both paths respectevily

File: docs/swagger.json (hunk 6 above)
Comment (frank-baier) [reply docs/swagger.json]: yes your first assumption is correct:

step 1. Upload document, which returns a token
step 2. Upload document ....
 ... when all documents are uploaded:
step n. Make policy request, including tokens for documents so you know what to copy in background.

File: docs/swagger.json (hunk 7 above)
Comment (frank-baier) [reply docs/swagger.json]: accepted, we will provide you with a mapping table next week

Comment (frank-baier) [reply docs/swagger.json]: around 20.03.

File: docs/swagger.json (hunk 8 above)
Comment (frank-baier) [reply docs/swagger.json]: accepted

File: docs/swagger.json (hunk 9 above)
Comment (frank-baier) [reply docs/swagger.json]: should be answered with aboves comment

File: docs/swagger.json (hunk 5 above)
Comment (frank-baier) [reply docs/swagger.json]: accepted, removed

File: docs/swagger.json (hunk 4 above)
Comment (frank-baier) [reply docs/swagger.json]: removed due to business decisions

File: .idea/encodings.xml [hunk 10]
```diff
@@ -0,0 +1,6 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
Comment (joachimprinzbach) [inline .idea/encodings.xml]: Why is this file added to git if the .idea folder is gitignored?

File: README.md [hunk 11]
```diff
@@ -6,4 +6,29 @@
 [![Codecov](https://img.shields.io/codecov/c/github/baloise/corellia.svg)](https://codecov.io/gh/baloise/corellia)
 [![gitpod-IDE](https://img.shields.io/badge/open--IDE-as--gitpod-blue.svg?style=flat&label=openIDE)](https://gitpod.io#https://github.com/baloise/corellia)
 
+## about
+
+The project hosts a classical b2b-interface (REST) specification for (insurance) contract data and document exchange.
```
Comment (joachimprinzbach) [inline README.md]: Probably should be called "Web-Api" instead of REST

File: pom.xml [hunk 12]
```diff
@@ -26,7 +26,7 @@
   <groupId>com.baloise.open.corellia</groupId>
```
Comment (joachimprinzbach) [inline pom.xml]: Having the same group and artifact id is considered a bad practice

File: pom.xml [hunk 13]
```diff
@@ -46,14 +46,60 @@
     <project.build.sourceEncoding>UTF-8</project.build.sourceEncoding>
     <maven.compiler.source>1.8</maven.compiler.source>
     <maven.compiler.target>1.8</maven.compiler.target>
-  
+
+    <swagger-version>2.0.1</swagger-version>
+    <jackson-version>2.9.1</jackson-version>
+    <jetty-version>9.4.9.v20180320</jetty-version>
```
Comment (joachimprinzbach) [inline pom.xml]: property naming of versions is inconsisted, stick to using dashes (-) or dots (.) - but do not mix

File: README.md (hunk 11 above)
Comment (frank-baier) [reply README.md]: accepted

File: pom.xml (hunk 12 above)
Comment (frank-baier) [reply pom.xml]: accepted

File: pom.xml [hunk 14]
```diff
@@ -46,14 +46,60 @@
     <project.build.sourceEncoding>UTF-8</project.build.sourceEncoding>
     <maven.compiler.source>1.8</maven.compiler.source>
     <maven.compiler.target>1.8</maven.compiler.target>
-  
+
+    <swagger-version>2.0.1</swagger-version>
```
[the remark this answers is not in the document]
Comment (frank-baier) [reply pom.xml]: declined: swagger version 2.0.x is correct. Swagger version 2 resolves to OpenAPI v3

File: pom.xml (hunk 13 above)
Comment (frank-baier) [reply pom.xml]: accepted

File: .idea/encodings.xml (hunk 10 above)
Comment (frank-baier) [reply .idea/encodings.xml]: interesting ... will try to remove

File: pom.xml (hunk 14 above)
[the remark this answers is not in the document]
Comment (joachimprinzbach) [reply pom.xml]: good to know - thx for the clarification

File: docs/openapi.json [hunk 15]
```diff
@@ -148,11 +149,38 @@
             "minLength" : 0,
             "type" : "string"
           },
+          "contactSalutation" : {
+            "maxLength" : 20,
```
Comment (joachimprinzbach) [inline docs/openapi.json]: do length restrictions make sense when using an enum?
